build: Use git archive as source tarball

hebasto opened the pull request:
This PR:
- is an alternative to #17104
- closes #16734
- closes #6753

The idea is clear described by some developers:
- [MarcoFalke](https://github.com/bitcoin/bitcoin/pull/17097#issuecomment-540691850):
> This whole concept of explicitly listing each and every file manually (or with a fragile wildcard) is an obvious sisyphean task. I'd say all we need to do is run git archive and be done with it forever, see #16734, #6753, #11530 ...

- [laanwj](https://github.com/bitcoin/bitcoin/pull/17097#issuecomment-540706025):
> I agree, I've never been a fan of it. I don't think we have any files in the git repository we don't want to ship in the source tarball.

---

The suggested changes have a downside which is pointed by [**luke-jr**](https://github.com/bitcoin/bitcoin/pull/17104#issuecomment-540828045): 
> ... but the distfile needs to include autogen-generated files.

This means that a user is not able to run `./configure && make` right away. One must run `./autogen.sh` at first.

Here are opinions about mandatory use of `./autogen.sh`:
- [ryanofsky](https://github.com/bitcoin/bitcoin/issues/16734#issuecomment-534139356):
> It's probably ok to require autogen. I think historically configure scripts were supposed to work on obscure unix systems that would just have a generic shell + make tool + c compiler, and not necessarily need gnu packages like m4 which are needed for autogen.

- [laanwj](https://github.com/bitcoin/bitcoin/issues/16734#issuecomment-540729483):
> I also think it's fine to require autogen. What is one dependency more, if you're building from source.

---

~Also this PR provides Windows users with ZIP archives of the sources. Additionally the commit ID is stored in these ZIP files as a file comment:~

---

Note for reviewers: please verify is `git archive` output deterministic?

hebasto committed:
build: Use git archive as source tarball

hebasto committed:
build: Drop SOURCEDIST reordering

Making SOURCEDIST deterministic is needless since a git archive is used
as the source tarball.

laanwj commented:
Thank you for picking this up.

laanwj commented on contrib/gitian-descriptors/gitian-linux.yml:
```diff
@@ -147,13 +147,6 @@ script: |
-  mkdir -p temp
-  pushd temp
-  tar -xf ../$SOURCEDIST
-  find bitcoin-* | sort | tar --mtime="$REFERENCE_DATETIME" --no-recursion --mode='u+rw,go+r-w,a+X' --owner=0 --group=0 -c -T - | gzip -9n > ../$SOURCEDIST
```
So this is not needed for the `git archive` output?
That's nice!

hebasto replied:
>    * The timestamp seems to be the same for all the files, but I don't know where it comes from. Oh it's the "CommitDate" of the last commit!

Yes. From [`git archive` docs](https://git-scm.com/docs/git-archive):
> `git archive` behaves differently when given a tree ID versus when given a **commit ID** or tag ID. In the first case the current time is used as the modification time of each file in the archive. In the latter case **the commit time as recorded in the referenced commit object is used** instead.

laanwj replied:
Seems this is the same for the `mtime` in the global header.

laanwj commented:
> Also this PR provides Windows users with ZIP archives of the sources. Additionally the commit ID is stored in these ZIP files as a file comment:

It looks like the generated `.tar` archive also has a comment embedded with the commit ID. I do not know any way (besides hexdump) of getting this information.

```
00000200  35 32 20 63 6f 6d 6d 65  6e 74 3d 32 34 61 37 32  |52 comment=24a72|
00000210  37 65 32 33 65 34 31 34  33 62 63 63 34 65 35 64  |7e23e4143bcc4e5d|
00000220  66 61 63 35 33 36 62 61  65 38 64 38 32 36 31 64  |fac536bae8d8261d|
00000230  33 32 61 0a 00 00 00 00  00 00 00 00 00 00 00 00  |32a.............|
00000240  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
```

hebasto commented:
> It looks like the generated `.tar` archive also has a comment embedded with the commit ID. I do not know any way (besides hexdump) of getting this information.

Yes. From [`git archive` docs](https://git-scm.com/docs/git-archive):
> Additionally the commit ID is stored in a global extended pax header if the tar format is used; it can be extracted using `git get-tar-commit-id`.

MarcoFalke commented:
ACK 86d9fae2943ad0d65ced2737f205db33f0bdc324 🤞

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 86d9fae2943ad0d65ced2737f205db33f0bdc324 🤞
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgJ3Av+JDgUre1dSaM2RmQQMxS3Qs1MeW1A2byPrGwosk53SkUWU/Pbxls2Bnl4
W2yF7J5HV8C95/864vUi9b1GRpE5Q3H/khgD1jqXUTXZW1VfQ8OtDImh9K8kNqf0
JvtucQyMoyiZXN1Enh+ck4/BRhEOD1aSGHQHW0lPpSDTZfh2oEhHA5dpzSYpOv+K
/pJxHHNNVGBaGaVuwVbyhAAievXySr9uS86w/q0e0NVnS4k2Jpac/e80hNVP6YK0
U4FgWCmoAl9Zxoc73YkBSPZpoajrN8vqbH7/DUIFR8tzBTIiAl3/yQI6gK97JH/r
v1HJ9gDgqa5euZncu536jsHwOhh6KWIC7D1XBbEVKL1k1Tr3gioI6F6Su1A/2OTR
75RQGFb99MxRONLI+Kf4iy34hG//8EE/R6yRMRNp28Lxx4nUp+i0ZhgxSwOrsiz7
Qlt14ejMnYC728yED74nsqNgVPqZbp5n5yn3M3G9Ip5/Xchc542UwuOuwLyluNgl
ZqFZrTHX
=63lK
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `70fc07cf6b63dd0d692f649fe3b0e90da30ba9c76c335ff9c9c3897f09d7551b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010870fc07cf6b63dd0d692f649fe3b0e90da30ba9c76c335ff9c9c3897f09d7551bf010e9bf561a35c3e5f884d799dc3f7cb5d008fff0101889b22d912b0a250a54bbe51b677c1108f1045e6d03a0f008ffa7dc3b345d74df0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f0109f1c2c3f2050c5f7864eb43f822342a108f020d79d0c8e7b7db1c5f04d2eab0d6a65c235a615f1439fe23b8088dbe84cc15f1a08f1045e6d03a1f00861ad4911991bb89e0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

MarcoFalke commented on contrib/gitian-descriptors/gitian-win.yml:
```diff
@@ -156,7 +156,7 @@ script: |
 
   mkdir -p ${OUTDIR}/src
-  git archive --output=${OUTDIR}/src/${DISTNAME}.tar.gz HEAD
+  git archive --output=${OUTDIR}/src/${DISTNAME}.zip HEAD
```
Why?

The tar is provided so that users can compile from source (without having to clone from GitHub and checkout the tag). I don't think we really support compiling on Windows. And when we do, the extra dependency to extract a tar file is probably not too much to ask for from a windows user.

Finally, as pointed out by @laanwj, this probably does need adjustment on the Website. Not sure if it is worth it.

hebasto replied:
Commit dropped (https://github.com/bitcoin/bitcoin/pull/18331#issuecomment-599098374).

hebasto commented:
Updated f05f8a49e02b49b6bcbdb82681b4c67509e8db73 -> 86d9fae2943ad0d65ced2737f205db33f0bdc324 ([pr18331.01](https://github.com/hebasto/bitcoin/commits/pr18331.01) -> [pr18331.02](https://github.com/hebasto/bitcoin/commits/pr18331.02), [compare](https://github.com/hebasto/bitcoin/compare/pr18331.01..pr18331.02)):

- dropped "build: Suggest source as zip archive for Windows" commit

MarcoFalke commented on contrib/gitian-descriptors/gitian-linux.yml:
```diff
@@ -196,5 +189,6 @@ script: |
     cd ../../
```
3 lines up:


```
+ cp ../README.md bitcoin-0.19.99/
cp: cannot stat '../README.md': No such file or directory

hebasto replied:
Fixed: https://github.com/bitcoin/bitcoin/pull/18331#issuecomment-599194633

hebasto committed:
build: Drop needless EXTRA_DIST content

Some EXTRA_DIST content is needless since a git archive is used as the
source tarball.

hebasto commented:
Updated 86d9fae2943ad0d65ced2737f205db33f0bdc324 -> c3c8d3cb06c72e14b5e264ca1492ce8a758814c7 ([pr18331.02](https://github.com/hebasto/bitcoin/commits/pr18331.02) -> [pr18331.03](https://github.com/hebasto/bitcoin/commits/pr18331.03), [compare](https://github.com/hebasto/bitcoin/compare/pr18331.02..pr18331.03)):

- reworked docs processing

hebasto commented:
Updated c3c8d3cb06c72e14b5e264ca1492ce8a758814c7 -> e4d366788bc2e8dce8e6ca572fce08d913d15d6b ([pr18331.03](https://github.com/hebasto/bitcoin/commits/pr18331.03) -> [pr18331.04](https://github.com/hebasto/bitcoin/commits/pr18331.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18331.03..pr18331.04); diff [18331.02..pr18331.04](https://github.com/hebasto/bitcoin/compare/pr18331.02..pr18331.04)):

- fixed windows installer (was broken on the previous push)

MarcoFalke commented:
For archival purposes, since DrahtBot wipes the builds after a while:

[bitcoin-0.19.99.tar.gz](https://github.com/bitcoin/bitcoin/files/4338567/bitcoin-0.19.99.tar.gz)

[bitcoin-0.19.99_new.tar.gz](https://github.com/bitcoin/bitcoin/files/4338570/bitcoin-0.19.99_new.tar.gz)



re-ACK e4d366788bc2e8dce8e6ca572fce08d913d15d6b, only change is adding two dots in a the path 🛳

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK e4d366788bc2e8dce8e6ca572fce08d913d15d6b, only change is adding two dots in a the path 🛳
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUirRgv/YnwsyN8A8pf5eyxqgi5qb9voP10DH/iI0wI47kkRfgNxHNuJE9hWIQMR
8DSE3rwNHWr9R1cJrJqV6Gy4xeLXsmE6VGsW1VXc8NeKg8D+UZAgDkMJ/onVPilb
RNp8sVRluXvr+NO+rh3Id3gYkz5MNHm9MIz+Na9UZYEDTlByed8DhJq6IhmqCFjc
gcdU4iPFNkJG+grPRR5++hFy4PcgOjIbHpThuZPzqaxM4hK83Clbucj0ICAIviRG
zdSLm/i9YXzQZNYHDEIHGePVChJ8iYGgpSNIRo2z3m1XsNqGQvcG6wdGMkF1K8Ju
k+Yazrr6aWy+mxwJRb621xUVCQlJrBfVE0/xTNT5uioBveJrQR95eTrP20oV0GjI
lmisdFa03kmpiHkHWBQ5wMFslITPNe/4enAvukOlMvexsuHZig2sBEB9EoG+Gfzc
WGjXGekdXal3RLhrm+tsNCjeyrUkFaFAaMDg4TTEVFHfkXNyI9Yp5gOVwS5Pr9IP
Fsh+65rN
=lSr7
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `885ed0aba5f83eaf085d59c35d2e5f2c7257ad642ea0929b4023d0c90888c9b0  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108885ed0aba5f83eaf085d59c35d2e5f2c7257ad642ea0929b4023d0c90888c9b0f0102486208e60f5eef6ecd956e073fe802808fff01004a34d9bf16802fed7c54e9c1c7a524708f120c01402521bfc59cfcd7f7805c10711077e2ef4f3b64f93e23b120c3ba367df2f08f1045e6f8d17f008d9261997a7e493b20083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01042aac25ea818d6b8d41b81f57a4fab8008f020b73ec9f1481d9205083b7b3e07dba94437bce309019b7ed9d943f418fc7a6e8408f1045e6f8d17f008b91a2138423ffedb0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f0109600f13ef9b1bb9746d4b7bea3e5f59908f0205f490b6ccfd91362e6a007240cecf77aac8ea4789bd255d2342ab3ce63ee94de08f1045e6f8d17f008e28c6cf73f5505360083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented:
Why is `GIT_ARCHIVE` set in `src/clientversion.cpp` even on current master? I thought we don't use `git archive` yet?!

MarcoFalke commented:
To answer my own question, we do use `git archive`:

```
$ git grep -W 'GIT) archive'
Makefile.am=dist-hook:
Makefile.am:    -$(GIT) archive --format=tar HEAD -- src/clientversion.cpp | $(AMTAR) -C $(top_distdir) -xf -

MarcoFalke commented:
@fanquake @laanwj Anything left to do here? I think this is ready

hebasto commented:
@MarcoFalke 
> Why is `GIT_ARCHIVE` set in `src/clientversion.cpp` even on current master? I thought we don't use `git archive` yet?!

This is the way to define `GIT_ARCHIVE`, `GIT_COMMIT_ID`, and `GIT_COMMIT_DATE` macros:
https://github.com/bitcoin/bitcoin/blob/86623873095f8d73fd28ad323ed3d06e20433176/src/clientversion.cpp#L44-L48

using git `export-subst` attribute.

laanwj commented:
ACK e4d366788bc2e8dce8e6ca572fce08d913d15d6b

theuni commented:
Sorry I got to this too late, but... I think this is a bad idea.

Requiring autogen essentially defeats the point of using autotools, which is to generate a portable configure script. This will likely be a painful move for downstream maintainers and packagers, who will now have to adjust their build dependencies.

Enumerating each file seems like a perfectly reasonable ask for a project with precision requirements as this one. Imo this is a move in the wrong direction.

ryanofsky commented:
> Requiring autogen essentially defeats the point of using autotools

This seems like a real concern. Does this PR now require running autogen when it wasn't required before? If so, it seems like it should have release notes and documentation updates.

MarcoFalke commented:
> painful move for downstream maintainers and packagers, who will now have to adjust their build dependencies.

For a major version bump they'd need to adjust the build dependencies anyway (e.g. remove openssl and boost chrono, ...). I don't really see how adding autotools is going to be painful. Though, if it really is that painful, we could run it in the gitian vm and append the resulting configure script to the archive.

> it should have release notes and documentation updates.

We don't really have any documentation about our release tarball, so there is nothing to update I think. I am happy to leave a note in the release notes.

ryanofsky commented:
> We don't really have any documentation about our release tarball, so there is nothing to update I think. I am happy to leave a note in the release notes.

I probably should familiarize myself with documentation more. I wrote my comment above assuming this PR required a new install step that was optional before, so installation instructions would be updated.

I don't think we should hesitate to write release notes for user visible changes. PR descriptions are often unhelpful because they focus on implementation details, while release notes force you to summarzie what the actual effect of a change is (it's still unclear here what files this PR removes from the tarball).

theuni commented:
> For a major version bump they'd need to adjust the build dependencies anyway (e.g. remove openssl and boost chrono, ...). I don't really see how adding autotools is going to be painful.

I'm not complaining about forcing downstreams to update. I'm complaining about the _contents_ of that update. They're going to have to add a pile of new build-time dependencies (autotools dev packages), which may or may not pose further complications.

hebasto commented:
Release notes added in #18439.

MarcoFalke commented:
@ryanofsky Our only instruction of building are here: https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md#to-build . The first step for them is ./autogen.sh (so nothing changed for them). There is also a source code download link on https://bitcoincore.org/en/download/ . This is what is going to change after this pull request. None of our own packaging scripts should change as an effect of that. They either use the static binaries or the git archive already https://github.com/bitcoin-core/packaging/blob/2c1d6a4cc67411beb3e035d126b56bef3cf7e1f1/debian/rules#L30

I can't comment on how hard it will be for other third party package maintainers to run autogen, but at least the debian one has a full copy of the git repo: https://sources.debian.org/src/bitcoin/0.18.1%7Edfsg-1/ . So I am not sure if there is anyone who depends on this targz and depends on this targz having the configure script included. My opinion is to wait for complaints before changing things again...

ryanofsky commented:
> I'm not complaining about forcing downstreams to update. I'm complaining about the _contents_ of that update. They're going to have to add a pile of new build-time dependencies (autotools dev packages), which may or may not pose further complications.

Are there particular operating systems / distros you're concerned about here? I haven't had the most autotools experience in the world, but I haven't personally run into cases where earlier `autoreconf` steps fail, it's always the later config / make stages that seem more fragile.

Or maybe there are systems where autotools and m4 are flat out broken or unavailable?

theuni commented:
I'm not aware of any. Practically, this will likely simply mean that maintainers add deps to a list and move on.

But at a higher level, we can offer a pre-processed source tarball which eases the burden on the builder (as well as adding improving determinism by pegging build tools at tag time) or we could not.

After this change, our source tarball is simply a shortcut to a git clone and checkout. Maybe that's ok, but it just doesn't seem worth the hassle of finding out to me.

But sure, now that it's in, we can just wait and see if it causes any real-world issues at release time.

MarcoFalke commented:
We couldn't keep up listing all needed files (or even establish guidelines about what files are needed) reliably, so this was the primary reason to change the source tarball. If the missing configure script is a burden or causes issues otherwise, what about appending it to the git archive after running autogen in the gitian vm?

This seems to solve both concerns?

theuni commented:
> We couldn't keep up listing all needed files (or even establish guidelines about what files are needed) reliably, so this was the primary reason to change the source tarball. If the missing configure script is a burden or causes issues otherwise, what about appending it to the git archive after running autogen in the gitian vm?
> 
> This seems to solve both concerns?

This is exactly what gitian did before this change. It's more than just 'configure' though, 'make dist' appends all bootstrapped files to the source tarball.

It sounds like you would just want an additional step of adding missing git files to that tarball.
